Clean up after unexpectedly terminated build

Honny1 opened the pull request:
The `podman system prune` command can remove build containers that were created during the build but were not removed because the build terminated unexpectedly. 

By default, build containers are not removed to prevent interference with builds in progress. Use the **--build** flag when running the command to remove build containers as well.

### Reproducer:
- Containerfile:
```Dockerfile
FROM ubi8/ubi
RUN truncate -s 10G out
RUN echo "Hi"
RUN sleep infinity
```
- Test script `run.sh`:
```bash
#!/usr/bin/env bash

podman build -f Containerfile -t podmanleaker &
sleep 60 && kill -9 $! 
```
- measure the size of current images, containers, etc... Before build
```bash
podman unshare du -sh ~/.local/share/containers/
```
- Test script (Note: requires disk space of about 32 GB)
```bash
./run.sh
```
- measure the size of current images, containers, etc... After termination of build
```bash
podman unshare du -sh ~/.local/share/containers/
```
- Clean up leftovers after build
```bash
podman system prune --build -f
```
- measure the size of current images, containers, etc... After cleanup build
```bash
podman unshare du -sh ~/.local/share/containers/
```
> The size should be the same as the first measurement but could be different if a base image is present in the system. 

Fixes: https://github.com/containers/podman/issues/14523
Fixes: https://github.com/containers/podman/issues/23683
Fixes: https://issues.redhat.com/browse/RHEL-62009

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
The `podman system prune` command now supports removing build containers with the new `--build` option. 
```

Luap99 commented on libpod/runtime.go:
```diff
@@ -1264,6 +1265,45 @@ func (r *Runtime) LockConflicts() (map[uint32][]string, []uint32, error) {
+		if err != nil {
+			return stageContainersPruneReports, err
+		}
+		if _, err := os.Stat(filepath.Join(path, "buildah.json")); errors.Is(err, fs.ErrNotExist) {
```
use fileutils.Exists() 

@nalind Is there another (better?) way to check if a storage container is from buildah?

Luap99 commented on libpod/runtime.go:
```diff
@@ -1264,6 +1265,45 @@ func (r *Runtime) LockConflicts() (map[uint32][]string, []uint32, error) {
+		size, err := r.store.ContainerSize(container.ID)
+		if err != nil {
+			report.Err = err
+			logrus.Warnf("Failed to get size of build stage container %s: %v", container.ID, err)
```
do not report the error to the caller and also print a warning.
Just reporting the errors back to the caller is good enough. The caller can then log once.

Luap99 commented on libpod/runtime.go:
```diff
@@ -1264,6 +1265,45 @@ func (r *Runtime) LockConflicts() (map[uint32][]string, []uint32, error) {
+		report.Size = uint64(size)
+
+		if err := r.store.DeleteContainer(container.ID); err != nil {
+			report.Err = err
+			logrus.Warnf("Failed to remove build stage container %s: %v", container.ID, err)
```
same here

Luap99 commented on pkg/domain/infra/abi/system.go:
```diff
@@ -61,21 +61,70 @@ func (ic *ContainerEngine) Info(ctx context.Context) (*define.Info, error) {
+
+	reclaimedSpace := (uint64)(0)
+	found := true
+	for found {
```
Why is this a loop? Running once should be enough.

Luap99 commented on pkg/domain/infra/abi/system.go:
```diff
@@ -61,21 +61,70 @@ func (ic *ContainerEngine) Info(ctx context.Context) (*define.Info, error) {
+		found = false
+
+		// Prune Stage Containers
+		stageContainersPruneReports, err := ic.Libpod.PruneStageContainers()
+		if err != nil {
+			return nil, err
+		}
+		if len(stageContainersPruneReports) > 0 {
+			found = true
+		}
+		reclaimedSpace += reports.PruneReportsSize(stageContainersPruneReports)
+		systemPruneReport.ContainerPruneReports = append(systemPruneReport.ContainerPruneReports, stageContainersPruneReports...)
+
+		// Prune Images
+		imagePruneOptions := entities.ImagePruneOptions{
+			External:   true,
+			BuildCache: true,
+		}
+		imageEngine := ImageEngine{Libpod: ic.Libpod}
+		imagePruneReports, err := imageEngine.Prune(ctx, imagePruneOptions)
+		if err != nil {
+			return nil, err
+		}
+		if len(imagePruneReports) > 0 {
+			found = true
+		}
+		reclaimedSpace += reports.PruneReportsSize(imagePruneReports)
+		systemPruneReport.ImagePruneReports = append(systemPruneReport.ImagePruneReports, imagePruneReports...)
```
This seems inconsistent with the documentation, from the docs you wrote it sounds like it only prune the build containers but this prunes everything other containers and images as well.
I don't think this is desirable.
If we want that behavior then the code should not cause a conflict with --build option and only do this in addition to the existing code in SystemPrune() instead of duplicating the image logic here.

Luap99 commented on test/e2e/prune_test.go:
```diff
@@ -580,4 +588,69 @@ var _ = Describe("Podman prune", func() {
+
+		build := podmanTest.Podman([]string{"build", "-f", containerFilePath, "-t", "podmanleaker"})
+		// Build will never finish so let's kill it after 8 seconds to simulate a failed build that leaves stage containers
+		time.Sleep(8 * time.Second)
```
Why do we wait 8s? That slows down the test a lot, 1/2 s should be more than enough?!

Luap99 commented on test/e2e/prune_test.go:
```diff
@@ -580,4 +588,69 @@ var _ = Describe("Podman prune", func() {
+		create.WaitWithDefaultTimeout()
+		Expect(create).Should(ExitCleanly())
+
+		containerFilePath := filepath.Join(podmanTest.TempDir, "ContainerFile-"+stringid.GenerateRandomID())
```
there is really no need for a random file name

Luap99 commented on test/e2e/prune_test.go:
```diff
@@ -580,4 +588,69 @@ var _ = Describe("Podman prune", func() {
+		none := podmanTest.Podman([]string{"images", "-a"})
+		none.WaitWithDefaultTimeout()
+		Expect(none).Should(ExitCleanly())
+		hasNone, result := none.GrepString("<none>")
+		Expect(result).To(HaveLen(1))
+		Expect(hasNone).To(BeTrue())
+
```
This leads to horrible Expected false to be true errors.

Please use the proper matchers, something like `Expect(none.OutputToString()).To(ContanSubstring("none"))`

Luap99 commented on test/e2e/prune_test.go:
```diff
@@ -580,4 +588,69 @@ var _ = Describe("Podman prune", func() {
+		Expect(hasNone).To(BeTrue())
+
+		// have: volatile-containers.json, containers.json, containers.lock and container dir and 2 stage container dirs
+		dirents, err := os.ReadDir(containerStorageDir)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(dirents).To(HaveLen(6))
```
Does this matter? This peaks at rather internal storage details which can change requiring the test to be updated often. Would it not be better to run `buildah containers` instead? Then in the end ensure it is removed there?

Luap99 commented on test/e2e/prune_test.go:
```diff
@@ -580,4 +588,69 @@ var _ = Describe("Podman prune", func() {
+		after := podmanTest.Podman([]string{"images", "-a"})
+		after.WaitWithDefaultTimeout()
+		Expect(after).Should(ExitCleanly())
+		Expect(len(after.OutputToStringArray())).To(BeNumerically(">", 1))
```
Not sure what this should check, AFAIK in the test/e2e setup we will always have extra images from the additional store shown.

Luap99 commented on test/e2e/prune_test.go:
```diff
@@ -580,4 +588,69 @@ var _ = Describe("Podman prune", func() {
+		after.WaitWithDefaultTimeout()
+		Expect(after).Should(ExitCleanly())
+		Expect(len(after.OutputToStringArray())).To(BeNumerically(">", 1))
+		hasNoneAfter, result := after.GrepString("<none>")
+		Expect(result).To(BeEmpty())
+		Expect(hasNoneAfter).To(BeFalse())
+		hasNotLeakerImager, _ := after.GrepString("notleaker")
+		Expect(hasNotLeakerImager).To(BeTrue())
```
also needs to use proper matchers

Luap99 commented on test/e2e/prune_test.go:
```diff
@@ -580,4 +588,69 @@ var _ = Describe("Podman prune", func() {
+		hasNotLeakerImager, _ := after.GrepString("notleaker")
+		Expect(hasNotLeakerImager).To(BeTrue())
+
+		// still have: volatile-containers.json, containers.json, containers.lock and container dir
+		dirents, err = os.ReadDir(containerStorageDir)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(dirents).To(HaveLen(4))
```
same comment I would use `buildah containers` to check

Luap99 commented:
Also what this doesn't answer what happens if the build process is still running? And another thing that this does not cleanup is the networking, not sure if there is a good way for that but if you kill the build and run with bridge networking the interfaces/firewall rules and ipam db allocations are still leaked. They should be cleaned after reboot so not that bad like the storage leak and certainly not a blocker for this here. I just mention it as there are other leaks to consider too.

Honny1 commented:
@Luap99 I have incorporated your review. I've changed the approach and the `--build` flag allows the removal of build/stage containers for the `podman system prune` command, so it works the same as `--volumes`.

Luap99 commented on docs/source/markdown/podman-system-prune.1.md:
```diff
@@ -7,20 +7,28 @@ podman\-system\-prune - Remove all unused pods, containers, images, networks, an
+
+Removes any build containers that were created during the build, but were not removed because the build was unexpectedly terminated.
+
+> **This is not safe operation and should be executed only when no builds are in progress. It can interfere with builds in progress.**
```
I think the general style in the man pages was to write `Note: ...` or something like that. I don't think we have used `>` elsewhere. I don't mind it for the web view but I have not yet looked at how it looks in the rendered man page.

Luap99 commented on test/e2e/prune_test.go:
```diff
@@ -580,4 +586,57 @@ var _ = Describe("Podman prune", func() {
+		build := podmanTest.Podman([]string{"build", "-f", containerFilePath, "-t", "podmanleaker"})
+		// Build will never finish so let's wait for build to ask for SIGKILL to simulate a failed build that leaves stage containers
+		for {
+			if build.LineInOutputContains("Please use signal 9") {
+				break
+			}
```
this is great but maybe add a small sleep for like 10ms or something to not busy poll which also eats a lot of resources.

Second this needs a timeout, if the build process fails and never prints the output this would loop forever causing hard to debug issues in CI.

Luap99 reviewed (commented):
Thanks, code wise this seems good to me. just a minor comment on the man page and we need to make sure the test does not run forever

Luap99 commented on test/e2e/prune_test.go:
```diff
@@ -580,4 +587,64 @@ var _ = Describe("Podman prune", func() {
+		build := podmanTest.Podman([]string{"build", "-f", containerFilePath, "-t", "podmanleaker"})
+		// Build will never finish so let's wait for build to ask for SIGKILL to simulate a failed build that leaves stage containers.
+		// In case build never asks for SIGKILL, this will timeout and fail the test.
+		done := make(chan struct{})
+		go func() {
+			for {
+				if build.LineInOutputContains("Please use signal 9") {
+					break
+				}
+				time.Sleep(100 * time.Millisecond)
+			}
+			build.Signal(syscall.SIGKILL)
+			close(done)
+		}()
+		Eventually(done, defaultWaitTimeout).Should(BeClosed())
```
This will still hang the goroutine which means you leak the entire goroutine. This really needs a timeout in the loop so the loop actually ends.

```go
matchedOutput := false

for range 900 {
	if build.LineInOutputContains("Please use signal 9") {
		matchedOutput = true
		break
	}
	time.Sleep(100 * time.Millisecond)
}
if !matchedOutput {
	Fail("Did not match special string in podman build")
}
```

Also when using goroutines in a ginkgo test one should use `defer GinkgoRecover()` to ensure error reporting there works. In this case it may not strictly be needed as there is no actual matching done but in case of a panic it should still be recovered by ginkgo.

Honny1 replied:
Thanks. I didn't realize that. Fixed.

Luap99 reviewed (approved):
/lgtm

Luap99 commented:
Removed the lgtm label again (forgot there wasn't a second review)

rhatdan commented:
I am hitting this all the time when playing with AI builds

Thanks @Honny1 
/lgtm
